File: exercise/82/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if head is None:
            return None
        before = ListNode(0)
        before.next = head

        pre = before
        now = head

        remove = False

        while now:
            dup = now.next
            if dup is not None and now.val == dup.val:
                remove = True
                now.next = dup.next
            elif remove:
                pre.next = dup
                now = dup
                remove = False
            else:
                pre = pre.next
                now = now.next

        return head

File: exercise/82/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode dum(0);
        ListNode *pre, *first, *second;
        dum.next = head;
        pre = &dum, first = head;
        bool dup = false;
        while (first) {
            second = first->next;
            if (second && first->val == second->val) {
                first->next = second->next;
                delete second;
                second = first->next;
                dup = true;
            } else if (dup) {
                pre->next = second;
                delete first;
                first = pre->next;
            } else {
                pre = first;
                first = second;
            }
        }
        return dum.next;
    }
};
